core/src/digest/mod.rs

```diff
@@ -12,50 +9,67 @@ use sp_inherents::InherentData;
 // but WITHOUT ANY WARRANTY; without even the implied warranty of
 // MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 // GNU General Public License for more details.
-use sp_runtime::Digest;
+pub use aura::Digest as FudgeAuraDigest;
+pub use babe::Digest as FudgeBabeDigest;
+use sp_inherents::InherentData;
+use sp_runtime::{traits::Block, Digest, DigestItem};
+use thiserror::Error;
 
 mod aura;
 mod babe;
-use sp_runtime::traits::Block;
+
+#[derive(Error, Debug)]
+pub enum Error {
+	#[error("timestamp inherent data retrieval: {0}")]
+	TimestampInherentDataRetrieval(sp_inherents::Error),
+
+	#[error("timestamp inherent data not found")]
+	TimestampInherentDataNotFound,
+
+	#[error("babe inherent data retrieval: {0}")]
+	BabeInherentDataRetrieval(sp_inherents::Error),
+
+	#[error("babe inherent data not found")]
+	BabeInherentDataNotFound,
+}
 
 #[async_trait::async_trait]
 pub trait DigestCreator<B>
 where
 	B: Block,
 {
-	async fn create_digest(&self, parent: B::Header, inherents: InherentData)
-		-> Result<Digest, ()>;
+	async fn create_digest(&self, inherents: InherentData) -> Result<Digest, Error>;
 }
 
 #[async_trait::async_trait]
 pub trait DigestProvider<B: Block> {
-	async fn build_digest(
-		&self,
-		parent: &B::Header,
-		inherents: &InherentData,
-	) -> Result<Digest, ()>;
+	fn digest(&self, inherents: &InherentData) -> Result<DigestItem, Error>;
+
+	async fn build_digest(&self, inherents: &InherentData) -> Result<Digest, Error> {
+		Ok(Digest {
+			logs: vec![self.digest(inherents)?],
+		})
+	}
 
 	async fn append_digest(
 		&self,
 		digest: &mut Digest,
-		parent: &B::Header,
 		inherents: &InherentData,
-	) -> Result<(), ()>;
+	) -> Result<(), Error> {
+		digest.push(self.digest(inherents)?);
+		Ok(())
+	}
 }
 
 #[async_trait::async_trait]
 impl<F, Fut, B> DigestCreator<B> for F
 where
 	B: Block,
-	F: Fn(B::Header, InherentData) -> Fut + Sync + Send,
```

Comment:
This is a subsrate change? I would like to keep the header. I think some thinks need it in the Substarte world. Lets stay in sync with InherentCreator from substrate.

Reply:
Oh, I did not think about that and just removed it since we did not use it here so far. I re-added it now and adjusted everything else accordingly.